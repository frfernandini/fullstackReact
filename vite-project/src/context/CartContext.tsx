import { createContext, useContext, useState, type ReactNode, useEffect } from 'react';
import type { Producto } from '../components/Producto';

interface CartItem extends Producto {
    cantidad: number;
}

interface CartContextType {
    cartItems: CartItem[];
    addToCart: (product: Producto) => void;
    removeFromCart: (productId: string) => void;
    updateQuantity: (productId: string, newQuantity: number) => void;
    clearCart: () => void;
    cartCount: number;
}

const CartContext = createContext<CartContextType | undefined>(undefined);

export const useCart = () => {
    const context = useContext(CartContext);
    if (!context) {
        throw new Error('useCart must be used within a CartProvider');
    }
    return context;
};

export const CartProvider = ({ children }: { children: ReactNode }) => {
    const [cartItems, setCartItems] = useState<CartItem[]>(() => {
        const storedCart = localStorage.getItem('carrito');
        return storedCart ? JSON.parse(storedCart) : [];
    });

    useEffect(() => {
        localStorage.setItem('carrito', JSON.stringify(cartItems));
    }, [cartItems]);

    const addToCart = (product: Producto) => {
        setCartItems(prevItems => {
            const existingItem = prevItems.find(item => item.id === product.id);
            if (existingItem) {
                return prevItems.map(item =>
                    item.id === product.id ? { ...item, cantidad: item.cantidad + 1 } : item
                );
            }
            return [...prevItems, { ...product, cantidad: 1 }];
        });
        alert(`${product.titulo} agregado al carrito`);
    };

    const removeFromCart = (productId: string) => {
        setCartItems(prevItems => prevItems.filter(item => item.id !== productId));
    };

    const updateQuantity = (productId: string, newQuantity: number) => {
        if (newQuantity <= 0) {
            removeFromCart(productId);
        } else {
            setCartItems(prevItems =>
                prevItems.map(item =>
                    item.id === productId ? { ...item, cantidad: newQuantity } : item
                )
            );
        }
    };

    const clearCart = (mostrarConfirm: boolean = true) => {
        if (!mostrarConfirm || window.confirm("¿Estás seguro de que quieres vaciar el carrito?")) {
            setCartItems([]); // vacía el carrito
        }
    };

    const cartCount = cartItems.reduce((total, item) => total + item.cantidad, 0);

    return (
        <CartContext.Provider value={{ cartItems, addToCart, removeFromCart, updateQuantity, clearCart, cartCount }}>
            {children}
        </CartContext.Provider>
    );
};
