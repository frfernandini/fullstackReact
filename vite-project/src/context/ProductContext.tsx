import { createContext, useContext, useState, useEffect, type ReactNode } from 'react';
import type { Producto } from '../components/Producto';
import { productos as ProductosData } from '../components/ProductoData';

interface ProductContextType {
    products: Record<string, Producto>;
    addProduct: (product: Producto) => void;
    removeProduct: (productId: string) => void;
    clearProducts: () => void;
}

const ProductContext = createContext<ProductContextType | undefined>(undefined);

export const useProducts = () => {
    const context = useContext(ProductContext);
    if (!context) {
        throw new Error('useProducts must be used within a ProductProvider');
    }
    return context;
};

export const ProductProvider = ({ children }: { children: ReactNode }) => {
    const [products, setProducts] = useState<Record<string, Producto>>(() => {
        try {
            const storedProducts = localStorage.getItem('productos');
            if (storedProducts) {
                return JSON.parse(storedProducts);
            }
        } catch (error) {
            console.error("Error parsing products from localStorage", error);
        }
        // If nothing in localStorage or it's invalid, initialize with default data
        const initialProducts: Record<string, Producto> = {};
        ProductosData.forEach(p => {
            initialProducts[p.id] = p;
        });
        return initialProducts;
    });

    useEffect(() => {
        try {
            localStorage.setItem('productos', JSON.stringify(products));
        } catch (error) {
            console.error("Error saving products to localStorage", error);
        }
    }, [products]);

    const addProduct = (product: Producto) => {
        setProducts(prevProducts => ({
            ...prevProducts,
            [product.id]: product
        }));
    };

    const removeProduct = (productId: string) => {
        setProducts(prevProducts => {
            const newProducts = { ...prevProducts };
            delete newProducts[productId];
            return newProducts;
        });
    };

    const clearProducts = () => {
        setProducts({});
    };

    return (
        <ProductContext.Provider value={{ products, addProduct, removeProduct, clearProducts }}>
            {children}
        </ProductContext.Provider>
    );
};
