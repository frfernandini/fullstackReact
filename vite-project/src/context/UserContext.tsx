import React, { createContext, useContext, useState, useEffect, type ReactNode } from 'react';
import type { User } from '../components/User.ts';

interface UserContextType {
    users: User[];
    addUser: (user: Omit<User, 'fechaRegistro'>) => { success: boolean; message: string };
    removeUser: (email: string) => void;
}

const UserContext = createContext<UserContextType | undefined>(undefined);

export const useUsers = () => {
    const context = useContext(UserContext);
    if (!context) {
        throw new Error('useUsers must be used within a UserProvider');
    }
    return context;
};

interface UserProviderProps {
    children: ReactNode;
}

export const UserProvider: React.FC<UserProviderProps> = ({ children }) => {
    const [users, setUsers] = useState<User[]>(() => {
        try {
            const storedUsers = localStorage.getItem('users');
            return storedUsers ? JSON.parse(storedUsers) : [];
        } catch (error) {
            console.error("Failed to parse users from localStorage", error);
            return [];
        }
    });

    useEffect(() => {
        try {
            localStorage.setItem('users', JSON.stringify(users));
        } catch (error) {
            console.error("Failed to save users to localStorage", error);
        }
    }, [users]);

    const addUser = (user: Omit<User, 'fechaRegistro'>): { success: boolean; message: string } => {
        if (users.some(u => u.email === user.email)) {
            return { success: false, message: 'El correo electrónico ya está registrado.' };
        }
        const newUser: User = {
            ...user,
            fechaRegistro: new Date().toISOString(),
        };
        setUsers(prevUsers => [...prevUsers, newUser]);
        return { success: true, message: 'Usuario registrado con éxito.' };
    };

    const removeUser = (email: string) => {
        setUsers(prevUsers => prevUsers.filter(user => user.email !== email));
    };

    return (
        <UserContext.Provider value={{ users, addUser, removeUser }}>
            {children}
        </UserContext.Provider>
    );
};
