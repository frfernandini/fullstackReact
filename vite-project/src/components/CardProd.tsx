import React from 'react';
import { Link } from 'react-router-dom';
import type { Producto } from './Producto';
import { useCart } from '../context/CartContext';

type CardProdProps = {
  producto: Producto;
};

export const CardProd: React.FC<CardProdProps> = ({ producto }) => {
  const { addToCart } = useCart();
  const tieneOferta = producto.oferta && (producto.descuento ?? 0) > 0;

  const precioOriginal = producto.precio;
  const precioFinal = tieneOferta
    ? Math.round(precioOriginal - (precioOriginal * (producto.descuento ?? 0) / 100))
    : precioOriginal;

  const handleAddToCart = (e: React.MouseEvent) => {
    e.preventDefault();
    e.stopPropagation();
    addToCart(producto);
  };

  return (
    <Link to={`/productos/${producto.id}`} className="card-product" role="link">
      <img src={producto.imagen} alt={producto.titulo} />

      <div className="card-content">
        <h3 className="card-titulo">{producto.titulo}</h3>

        {!tieneOferta ? (
          <p className="card-precio">${precioOriginal.toLocaleString()}</p>
        ) : (
          <div className="card-precio">
            <p>
              <span className="text-muted text-decoration-line-through">${precioOriginal.toLocaleString()}</span>
              <br />
              <span className="fw-bold text-success">${precioFinal.toLocaleString()}</span>
            </p>
            <span className="badge bg-danger">-{producto.descuento}% OFF</span>
          </div>
        )}

        <p className="card-desc">{producto.descripcion}</p>
      </div>

      <button className="btn-agregar" onClick={handleAddToCart}>
        <i className="bi bi-cart-plus" /> Agregar al Carrito
      </button>
    </Link>
  );
};

export default CardProd;
