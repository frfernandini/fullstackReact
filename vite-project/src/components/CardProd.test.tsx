import { fireEvent, render, screen } from "@testing-library/react";
import { describe, expect, test, vi, beforeEach } from "vitest";
import { BrowserRouter } from "react-router-dom";
import { CardProd } from "./CardProd";
import { CartProvider } from "../context/CartContext";
import type { Producto } from "./Producto";

// Mock de producto sin oferta
const mockProductoSinOferta: Producto = {
    id: '1',
    titulo: 'Café Colombiano Premium',
    precio: 15000,
    descripcion: 'Café de alta calidad con notas frutales',
    categoria: 'Café',
    imagen: '/img/cafe-colombiano.jpg',
    oferta: false
};

// Mock de producto con oferta
const mockProductoConOferta: Producto = {
    id: '2',
    titulo: 'Té Verde Orgánico',
    precio: 12000,
    descripcion: 'Té verde natural con propiedades antioxidantes',
    categoria: 'Té',
    imagen: '/img/te-verde.jpg',
    oferta: true,
    descuento: 20
};

// Componente wrapper que provee el contexto y el router necesarios
const renderWithProviders = (producto: Producto) => {
    return render(
        <BrowserRouter>
            <CartProvider>
                <CardProd producto={producto} />
            </CartProvider>
        </BrowserRouter>
    );
};

describe('CardProd', () => {
    beforeEach(() => {
        // Simular window.alert para que no interrumpa los tests
        window.alert = vi.fn();
        window.confirm = vi.fn(() => true);
    });

    test("Debería renderizar correctamente un producto sin oferta", () => {
        renderWithProviders(mockProductoSinOferta);

        // Verificar que el título se muestre
        expect(screen.getByText('Café Colombiano Premium')).toBeDefined();
        
        // Verificar que el precio se muestre (con coma o punto según locale)
        expect(screen.getByText(/\$15[.,]000/)).toBeDefined();
        
        // Verificar que la descripción se muestre
        expect(screen.getByText('Café de alta calidad con notas frutales')).toBeDefined();
        
        // Verificar que el botón de agregar al carrito exista
        expect(screen.getByText(/Agregar al Carrito/i)).toBeDefined();
    });

    test("Debería renderizar correctamente un producto con oferta", () => {
        renderWithProviders(mockProductoConOferta);

        // Verificar que el título se muestre
        expect(screen.getByText('Té Verde Orgánico')).toBeDefined();
        
        // Verificar que el precio original (tachado) se muestre (con coma o punto según locale)
        expect(screen.getByText(/\$12[.,]000/)).toBeDefined();
        
        // Verificar que el precio con descuento se muestre (12000 - 20% = 9600)
        expect(screen.getByText(/\$9[.,]600/)).toBeDefined();
        
        // Verificar que el badge de descuento se muestre
        expect(screen.getByText('-20% OFF')).toBeDefined();
    });

    test("Debería llamar a addToCart cuando se hace clic en el botón", () => {
        renderWithProviders(mockProductoSinOferta);

        // Buscar el botón de agregar al carrito
        const btnAgregar = screen.getByText(/Agregar al Carrito/i);
        
        // Simular el clic
        fireEvent.click(btnAgregar);
        
        // Verificar que se haya llamado a window.alert (por el addToCart)
        expect(window.alert).toHaveBeenCalledWith('Café Colombiano Premium agregado al carrito');
    });

    test("Debería tener un link hacia la página de detalles del producto", () => {
        renderWithProviders(mockProductoSinOferta);

        // Buscar el link por su rol
        const link = screen.getByRole('link');
        
        // Verificar que el link tenga el href correcto
        expect(link.getAttribute('href')).toBe('/productos/1');
    });

    test("Snapshot: Debería mantener la estructura del componente sin oferta", () => {
        const { container } = renderWithProviders(mockProductoSinOferta);
        
        // Crear un snapshot del componente renderizado
        expect(container.firstChild).toMatchSnapshot();
    });

    test("Snapshot: Debería mantener la estructura del componente con oferta", () => {
        const { container } = renderWithProviders(mockProductoConOferta);
        
        // Crear un snapshot del componente renderizado
        expect(container.firstChild).toMatchSnapshot();
    });

    test("Debería calcular correctamente el precio con descuento", () => {
        // Producto con 30% de descuento
        const productoDescuento30: Producto = {
            id: '3',
            titulo: 'Producto Test',
            precio: 10000,
            descripcion: 'Descripción test',
            categoria: 'Test',
            imagen: '/img/test.jpg',
            oferta: true,
            descuento: 30
        };

        renderWithProviders(productoDescuento30);

        // Precio original: 10000
        // Descuento 30%: 10000 - 3000 = 7000
        // Usar regex para aceptar tanto punto como coma como separador de miles
        expect(screen.getByText(/\$7[.,]000/)).toBeDefined();
        expect(screen.getByText('-30% OFF')).toBeDefined();
    });

    test("No debería mostrar el badge de oferta si el producto no tiene oferta", () => {
        renderWithProviders(mockProductoSinOferta);

        // Verificar que el badge de oferta NO esté presente
        const badge = screen.queryByText(/-\d+% OFF/);
        expect(badge).toBeNull();
    });

    test("Debería mostrar la imagen del producto con el alt correcto", () => {
        renderWithProviders(mockProductoSinOferta);

        // Buscar la imagen por el alt
        const imagen = screen.getByAltText('Café Colombiano Premium');
        
        // Verificar que la imagen tenga el src correcto
        expect(imagen.getAttribute('src')).toBe('/img/cafe-colombiano.jpg');
    });
});
