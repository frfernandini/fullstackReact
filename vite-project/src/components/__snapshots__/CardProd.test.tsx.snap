// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CardProd > Snapshot: Debería mantener la estructura del componente con oferta 1`] = `
<a
  class="card-product"
  data-discover="true"
  href="/productos/2"
  role="link"
>
  <img
    alt="Té Verde Orgánico"
    src="/img/te-verde.jpg"
  />
  <div
    class="card-content"
  >
    <h3
      class="card-titulo"
    >
      Té Verde Orgánico
    </h3>
    <div
      class="card-precio"
    >
      <p>
        <span
          class="text-muted text-decoration-line-through"
        >
          $
          12,000
        </span>
        <br />
        <span
          class="fw-bold text-success"
        >
          $
          9,600
        </span>
      </p>
      <span
        class="badge bg-danger"
      >
        -
        20
        % OFF
      </span>
    </div>
    <p
      class="card-desc"
    >
      Té verde natural con propiedades antioxidantes
    </p>
  </div>
  <button
    class="btn-agregar"
  >
    <i
      class="bi bi-cart-plus"
    />
     Agregar al Carrito
  </button>
</a>
`;

exports[`CardProd > Snapshot: Debería mantener la estructura del componente sin oferta 1`] = `
<a
  class="card-product"
  data-discover="true"
  href="/productos/1"
  role="link"
>
  <img
    alt="Café Colombiano Premium"
    src="/img/cafe-colombiano.jpg"
  />
  <div
    class="card-content"
  >
    <h3
      class="card-titulo"
    >
      Café Colombiano Premium
    </h3>
    <p
      class="card-precio"
    >
      $
      15,000
    </p>
    <p
      class="card-desc"
    >
      Café de alta calidad con notas frutales
    </p>
  </div>
  <button
    class="btn-agregar"
  >
    <i
      class="bi bi-cart-plus"
    />
     Agregar al Carrito
  </button>
</a>
`;
